pkg/controller/nodes/task/k8s/event_watcher.go

```diff
@@ -0,0 +1,119 @@
+package k8s
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	eventsv1 "k8s.io/api/events/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/informers"
+	informerEventsv1 "k8s.io/client-go/informers/events/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/flyteorg/flytestdlib/logger"
+)
+
+type EventWatcher interface {
+	List(objectNsName types.NamespacedName, createdAfter time.Time) []*EventInfo
+}
+
+type eventWatcher struct {
+	informer    informerEventsv1.EventInformer
+	objectCache sync.Map
+}
+
+type objectEvents struct {
+	mu         sync.RWMutex
+	eventInfos map[types.NamespacedName]*EventInfo
+}
+
+// EventInfo stores detail about the event and the timestamp of the first occurrence.
+// All other fields are thrown away to conserve space.
+type EventInfo struct {
+	Reason     string
+	Note       string
+	CreatedAt  time.Time
+	RecordedAt time.Time
+}
+
+func (e *eventWatcher) OnAdd(obj interface{}) {
+	event := obj.(*eventsv1.Event)
+	objectNsName := types.NamespacedName{Namespace: event.Regarding.Namespace, Name: event.Regarding.Name}
+	eventNsName := types.NamespacedName{Namespace: event.Namespace, Name: event.Name}
+	v, _ := e.objectCache.LoadOrStore(objectNsName, &objectEvents{
+		eventInfos: map[types.NamespacedName]*EventInfo{},
+	})
+	objEvents := v.(*objectEvents)
+	objEvents.mu.Lock()
+	defer objEvents.mu.Unlock()
+	objEvents.eventInfos[eventNsName] = &EventInfo{
+		Reason:     event.Reason,
+		Note:       event.Note,
+		CreatedAt:  event.CreationTimestamp.Time,
+		RecordedAt: time.Now(),
+	}
+}
+
+func (e *eventWatcher) OnUpdate(_, newObj interface{}) {
+	// Dropping event updates since we only care about the creation
+}
+
+func (e *eventWatcher) OnDelete(obj interface{}) {
```

Comment:
can you briefly explain how k8s deletes events? is this done on object deletion? just want to make sure this cache does not grow unbounded.